completion_aggregator/api/v1/views.py

```diff
@@ -388,3 +390,109 @@ def get(self, request, course_key):
+
+        Example response:
+
+            TODO
+
+    """
+
+    def get(self, request, course_key):
+        """
+        Handler for GET requests
+        """
+        course_key = CourseKey.from_string(course_key)
+        paginator = self.pagination_class()  # pylint: disable=not-callable
+        requested_fields = self.get_requested_fields()
+        cohorts = UserCohorts(course_key)
+
+        course_cohorts = cohorts.get_course_cohorts()
+        if not course_cohorts:
+            raise NotFound()
+
+        enrollments = UserEnrollments().get_course_enrollments(course_key)
+        filtered_enrollments = self._filter_enrollments(
+            enrollments=enrollments, course_key=course_key, cohorts=cohorts)
+        paginated = paginator.paginate_queryset(
+            filtered_enrollments, self.request, view=self)
+        aggregator_queryset = self.get_queryset().filter(
+            course_key=course_key,
+            user__in=[enrollment.user for enrollment in paginated])
+        completions = aggregator_queryset.aggregate(
+            possible=Avg('possible'),
+            earned=Avg('earned'),
+            percent=Sum('earned') / Sum('possible'))
+        completions['course_key'] = course_key
+
+        # Return the paginated, serialized completions
+        serializer = self.get_serializer_class()(
+            instance=[completions],
+            requested_fields=requested_fields,
+            many=True
+        )
+
+        return paginator.get_paginated_response(serializer.data)
+
+    def _filter_enrollments(self, enrollments, course_key, cohorts):
+        """
+        Filter enrollments based on determined requirements
+        """
+        filtered_enrollments = []
+        roles_to_exclude = self.request.query_params.get('exclude_roles')
+        roles_to_exclude = roles_to_exclude.split(',')
+        cohort_filter = int(self.request.query_params.get('cohorts'))
+        for enrollment in enrollments:
+            enrolled_user = enrollment.user
+            if not UserEnrollments(enrolled_user).is_enrolled(course_key):
+                # Check if user has active enrollment in the course
+                # TODO ensure this check is necessary
+                # (probably users from the enrollment are active)
+                continue
+            if user_has_excluded_roles(
+                    enrolled_user, course_key, roles_to_exclude):
+                # User is a member of a role that should be excluded from
+                # the results.
+                continue
+            if 'staff' in roles_to_exclude and enrolled_user.is_staff:
+                # Double check in case staff role is handled by Django
+                # User is a staff member and staff needs to be excluded
+                continue
+            cohort_id = cohorts.get_user_cohorts(enrolled_user)
+            if cohort_id == cohort_filter:
+                # Keep user in the results
+                filtered_enrollments.append(enrollment)
```

Comment:
With tens of thousands of users in a course, this loop is going to be massively slow.  We need to filter the enrollments in the database.  Let me know if you need pointers on making more complex queries using the ORM.

Reply:
Maybe push the query refinements into methods that can be mocked.

Something like:
```python
aggregator_qs = self.get_queryset(course_key, enrollments)
if roles_to_exclude:
    self.exclude_roles(roles_to_exclude)
...

def get_initial_queryset(course_key, enrollments):
    return self.get_queryset().filter(course_key=course_key, user__in enrollment.user for enrollment in paginated_enrollments)

def exclude_roles(self, queryset, roles_to_exclude):
     if roles_to exclude:
        return queryset.exclude(user__courseaccessrole__role__in=roles_to_exclude)
...
```
And then mock get_initial_queryset to return a list, and mock exclude_users to filter that list.

Reply:
(Response made without looking at the latest updates to the code.  Checking it out now.